### PR TITLE
Python: Fix reading 0-bytes binary

### LINKED ISSUE #6435: PyIceberg: Avro decode EOF error
### Feature Request / Improvement

In reading manifests for a table for a table scan in PyIceberg 0.2.0, I get an EOFError.

Table was originally written in June 2022 with the most recent version of Spark/Iceberg at that time. (3.2.3 and 0.14.1 I think)

Full tracebook of the error:
```
Traceback (most recent call last):
  File "/read.py", line 16, in <module>
    print(tbl.scan(selected_fields=("time", "icao24","call_sign","origin_country")).to_arrow().to_pandas().head())
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/table/__init__.py", line 350, in to_arrow
    for task in self.plan_files():
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/table/__init__.py", line 335, in plan_files
    yield from (FileScanTask(file) for file in matching_partition_files)
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/table/__init__.py", line 335, in <genexpr>
    yield from (FileScanTask(file) for file in matching_partition_files)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/manifest.py", line 149, in <genexpr>
    return (entry.data_file for entry in live_entries(input_file))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/manifest.py", line 145, in <genexpr>
    return (entry for entry in read_manifest_entry(input_file) if entry.status != ManifestEntryStatus.DELETED)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/manifest.py", line 139, in read_manifest_entry
    for record in reader:
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/file.py", line 178, in __next__
    return self.__next__()
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/file.py", line 170, in __next__
    return next(self.block)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/file.py", line 106, in __next__
    return self.reader.read(self.block_decoder)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/reader.py", line 283, in read
    result[pos] = field.read(decoder)
                  ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/reader.py", line 283, in read
    result[pos] = field.read(decoder)
                  ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/reader.py", line 267, in read
    return self.option.read(decoder)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/reader.py", line 329, in read
    read_items[key] = self.value.read(decoder)
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/reader.py", line 234, in read
    return decoder.read_bytes()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/decoder.py", line 122, in read_bytes
    return self.read(self.read_int())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pyiceberg/avro/decoder.py", line 58, in read
    raise EOFError
EOFError
```

The table can be read successfully with Trino.

for interest, the schema of my table is 
```
table {
  1: time: optional long
  2: icao24: optional string
  3: callsign: optional string
  4: origin_country: optional string
  5: time_position: optional double
  6: last_contact: optional long
  7: longitude: optional double
  8: latitude: optional double
  9: baro_altitude: optional double
  10: on_ground: optional boolean
  11: velocity: optional double
  12: true_track: optional double
  13: vertical_rate: optional double
  14: geo_altitude: optional double
  15: squawk: optional string
  16: spi: optional boolean
  17: position_source: optional long
}
```

### Query engine

None

### PR DESCRIPTION
**Pull request by Fokko**
When a binary field would be zero bytes, we would still read the stream, causing a EOFErrors:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/manifest.py", line 139, in read_manifest_entry
    for record in reader:
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/file.py", line 194, in __next__
    return self.__next__()
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/file.py", line 186, in __next__
    return next(self.block)
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/file.py", line 113, in __next__
    return self.reader.read(self.block_decoder)
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/reader.py", line 260, in read
    struct.set(pos, field.read(decoder))  # later: pass reuse in here
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/reader.py", line 260, in read
    struct.set(pos, field.read(decoder))  # later: pass reuse in here
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/reader.py", line 233, in read
    return self.option.read(decoder)
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/reader.py", line 313, in read
    read_items[key] = self.value.read(decoder)
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/reader.py", line 200, in read
    return decoder.read_bytes()
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/decoder.py", line 118, in read_bytes
    return self.read(self.read_int())
  File "/Users/fokkodriesprong/Desktop/iceberg/python/pyiceberg/avro/decoder.py", line 54, in read
    raise EOFError
EOFError
```

Fixes #6435

Many thanks to @joshuarobinson for reporting this 🙏🏻

### CODE REVIEW

File: python/pyiceberg/avro/decoder.py [hunk 1]
```diff
@@ -50,10 +50,10 @@ def read(self, n: int) -> bytes:
             raise ValueError(f"Requested {n} bytes to read, expected positive integer.")
         read_bytes = self._input_stream.read(n)
         read_len = len(read_bytes)
-        if read_len <= 0:
-            raise EOFError
+        if read_len < 0:
+            raise EOFError(f"Got negative length: {read_len}")
```
[the remark this answers is not in the document]
Comment (rdblue) [reply python/pyiceberg/avro/decoder.py:54]: If I remember correctly, I think that this is how Python streams signal EOF. Now that we're using an Arrow stream, maybe that has changed?

[the remark this answers is not in the document]
Comment (Fokko) [reply python/pyiceberg/avro/decoder.py:54]: I'm not too opinionated about this, but it looks like a `ValueError` is not the most appropriate one:
image

File: python/pyiceberg/avro/decoder.py [hunk 2]
```diff
@@ -115,7 +115,8 @@ def read_bytes(self) -> bytes:
         """
         Bytes are encoded as a long followed by that many bytes of data.
         """
-        return self.read(self.read_int())
+        num_bytes = self.read_int()
+        return self.read(num_bytes) if num_bytes > 0 else b""
```
Comment (rdblue) [inline python/pyiceberg/avro/decoder.py:119]: Ah, this makes more sense! So we were requesting 0 bytes and hitting an `EOFError` because of it.

File: python/pyiceberg/avro/decoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply python/pyiceberg/avro/decoder.py:54]: Just to capture this somewhere, Python does use [0 bytes to signal EOF](https://docs.python.org/3/library/io.html#io.RawIOBase.read):

> If 0 bytes are returned, and size was not 0, this indicates end of file.

File: python/pyiceberg/avro/decoder.py (hunk 2 above)
Comment (rdblue) [reply python/pyiceberg/avro/decoder.py:119]: It would also make sense to put this behavior into `read(num_bytes)` since we could have other callers ask for 0 bytes by accident.

**Review by rdblue — APPROVED** [review]

File: python/pyiceberg/avro/decoder.py [hunk 3]
```diff
@@ -50,10 +50,10 @@ def read(self, n: int) -> bytes:
             raise ValueError(f"Requested {n} bytes to read, expected positive integer.")
         read_bytes = self._input_stream.read(n)
         read_len = len(read_bytes)
-        if read_len <= 0:
-            raise EOFError
+        if read_len < 0:
+            raise EOFError(f"Got negative length: {read_len}")
         elif read_len != n:
-            raise ValueError(f"Read {len(read_bytes)} bytes, expected {n} bytes")
+            raise EOFError(f"Read {read_len} bytes, expected {n} bytes")
```
[the remark this answers is not in the document]
Comment (rdblue) [reply python/pyiceberg/avro/decoder.py]: I don't think this is quite correct. If read_len is 0 then `EOFError` is correct. But it is perfectly fine for a read to return fewer bytes than requested. From [RawIOBase.read docs](https://docs.python.org/3/library/io.html#io.RawIOBase.read):

> Read up to size bytes from the object and return them. As a convenience, if size is unspecified or -1, all bytes until EOF are returned. Otherwise, only one system call is ever made. Fewer than size bytes may be returned if the operating system call returns fewer than size bytes.

If fewer bytes are returned this can either loop and call read until the required bytes are returned ("read fully"), or it can return a shorter array. Since this is a higher-level API, I think it should have read fully behavior.

File: python/pyiceberg/avro/decoder.py [hunk 4]
```diff
@@ -48,13 +48,18 @@ def read(self, n: int) -> bytes:
...
-        return read_bytes
+        data = b""
+
+        n_remaining = n
+        while n_remaining > 0:
+            data_read = self._input_stream.read(n_remaining)
+            read_len = len(data_read)
+            if read_len <= 0:
+                raise EOFError(f"Got negative length: {read_len}")
+            data += data_read
```
Comment (rdblue) [inline python/pyiceberg/avro/decoder.py:59]: Looks like this will reallocate and copy, according to [this post on efficient byte operations](https://www.guyrutenberg.com/2020/04/04/fast-bytes-concatenation-in-python/). From the post, it looks like we should either keep a list of read buffers and call join at the end, or use `bytearray` to accumulate data and convert at the end.

File: python/pyiceberg/avro/decoder.py [hunk 5]
```diff
@@ -48,13 +48,18 @@ def read(self, n: int) -> bytes:
...
-            raise ValueError(f"Read {len(read_bytes)} bytes, expected {n} bytes")
-        return read_bytes
+        data = b""
+
+        n_remaining = n
+        while n_remaining > 0:
+            data_read = self._input_stream.read(n_remaining)
+            read_len = len(data_read)
+            if read_len <= 0:
+                raise EOFError(f"Got negative length: {read_len}")
```
Comment (rdblue) [inline python/pyiceberg/avro/decoder.py:58]: This isn't necessarily negative. Most likely, it is going to be 0, so this would look strange. How about `f"EOF: read {read_len} bytes"`

File: python/tests/avro/test_decoder.py [hunk 6]
```diff
@@ -109,7 +109,7 @@ class OneByteAtATimeInputStream(InputStream):
 
     def read(self, size: int = 0) -> bytes:
         self.pos += 1
-        return int.to_bytes(1, self.pos, byteorder="little")
+        return self.pos.to_bytes(1, byteorder="little")
```
Comment (rdblue) [inline python/tests/avro/test_decoder.py:112]: Why was this needed?

File: python/pyiceberg/avro/decoder.py (hunk 4 above)
Comment (Fokko) [reply python/pyiceberg/avro/decoder.py:59]: A good call!

File: python/tests/avro/test_decoder.py (hunk 6 above)
Comment (Fokko) [reply python/tests/avro/test_decoder.py:112]: The test was actually incorrect. The `self.pos` is at the position of length, and the name (`OneByteAtATimeInputStream`) suggests that it only returns one byte at a time.

Without the change, the test fails:
```
    def test_read_single_byte_at_the_time() -> None:
        decoder = BinaryDecoder(OneByteAtATimeInputStream())
>       assert decoder.read(2) == b"\x01\x02"
E       AssertionError: assert bytearray(b'\x01\x01\x00') == b'\x01\x02'
E         At index 1 diff: 1 != 2
E         Left contains one more item: 0
E         Use -v to get more diff
```
It will return more bytes each time it is called:

image

image

image

File: python/pyiceberg/avro/decoder.py (hunk 4 above)
Comment (Fokko) [reply python/pyiceberg/avro/decoder.py:59]: I went with the list approach. It is slightly slower than the bytearray approach, but in our case we'll have to convert the `bytearray` back to bytes which also introduces a penalty:

```
```
